iothub/device/src/Pipeline/MqttTransportHandler.cs

```diff
@@ -1108,18 +1140,18 @@ private async Task HandleReceivedCloudToDeviceMessageAsync(MqttApplicationMessag
                     if (Logging.IsEnabled)
                         Logging.Error(this, "Cannot 'reject' or 'abandon' a received message over MQTT. Message will be acknowledged as 'complete' instead.");
                 }
-
-                // Note that MQTT does not support Abandon or Reject, so always Complete by acknowledging the message like this.
-                try
-                {
-                    await receivedEventArgs.AcknowledgeAsync(CancellationToken.None).ConfigureAwait(false);
-                }
-                catch (Exception ex)
+            }
+            else if(!_isDeviceReceiveMessageCallbackSet)
+            {
+                _messageQueue.Enqueue(receivedCloudToDeviceMessage);
```

Comment:
shouldn't the size be checked prior to enqueuing?

Reply:
since we are only keeping latest messages, we enqueue the new message and check if size > limit and then remove the oldest message.